### PR TITLE
Baseline LSTM implementation 

### PR DESCRIPTION
**Pull request by achyudhk**
## Baseline LSTM
Implementation of a standard LSTM using PyTorch and Torchtext for text classification baseline measurements

### Model Type
 - rand: All words are randomly initialized and then modified during training.
- static: A model with pre-trained vectors from [word2vec](https://code.google.com/archive/p/word2vec/). All words -- including the unknown ones that are initialized with zero -- are kept static and only the other parameters of the model are learned.
- non-static: Same as above but the pretrained vectors are fine-tuned for each task.

### Quick Start
 To run the model on Reuters dataset on static, just run the following from the Castor working directory.
 ```
python -m lstm_baseline --mode static
```

 ### Dataset
 We experiment the model on the following datasets.
 - Reuters dataset - ModApte splits

### Settings
 Adadelta is used for training.
 
### TODO
 - Add dataset results with different hyperparameters
- Parameters tuning

### CODE REVIEW

File: lstm_baseline/__main__.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    elif args.dataset == 'SST-2':
+        evaluate_dataset('dev', SST2, model, None, dev_iter, args.batch_size, args.gpu)
+        evaluate_dataset('test', SST2, model, None, test_iter, args.batch_size, args.gpu)
+    elif args.dataset == 'Reuters':
+        evaluate_dataset('dev', Reuters, model, None, dev_iter, args.batch_size, args.gpu)
+        evaluate_dataset('test', Reuters, model, None, test_iter, args.batch_size, args.gpu)
+    else:
+        raise ValueError('Unrecognized dataset')
+
+    if args.onnx:
```
Comment (Impavidity) [inline lstm_baseline/__main__.py]: Does the ONNX part work?
Do you test it?

File: lstm_baseline/model.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        if self.mode == 'rand':
+            x = self.embed(x) # (batch, sent_len, embed_dim)
+        elif self.mode == 'static':
+            x = self.static_embed(x)
+        elif self.mode == 'non-static':
+            x = self.non_static_embed(x)
+        else:
+            print("Unsupported Mode")
+            exit()
+        x, _ = self.lstm(x)
```
Comment (Impavidity) [inline lstm_baseline/model.py]: What's the value of `batch_first` option for the LSTM ? Is it False?
Then what's the input format of x?

Comment (achyudhk) [reply lstm_baseline/model.py]: Yeah, the default value is false. I anyway transpose the output to (batch, seq, feature) dimensions which is what the batch_first option does anyway.

Comment (achyudhk) [reply lstm_baseline/model.py]: I need to check the input dimensions again once Dragon is up tomorrow. I ported this code over to Castor and I might have missed something.

File: lstm_baseline/__main__.py (hunk 1 above)
Comment (achyudhk) [reply lstm_baseline/__main__.py]: Nope, it doesn't work. Throws a ONNX export failed (Couldn't export Python operator forward_flattened_wrapper) exception. Should I just remove the ONNX option for now?

Comment (Impavidity) [reply lstm_baseline/__main__.py]: yes please. And record this in readme.

Comment (achyudhk) [reply lstm_baseline/__main__.py]: Done.

**Comment by Impavidity** [comment]
One more suggestion, basically for the implementation of RNN model, we should provide length and use `rnn_utils.pack_padded_sequence`. This is a more reasonable implementation (do not means this will get better results). If you have extra time, you could compare these two implementations.

**Comment by achyudhk** [comment]
@Impavidity thanks for the tip. I added padding for packed sequences in #152 along with TensorboardX support. I didn't want to crowd this PR so made a new one. Also, I'll continue to push commits there so there is no need for a review right now.
